core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java

```diff
@@ -327,40 +342,82 @@ public static void writeOrder(Terms.Order order, StreamOutput out) throws IOExce
-                case TERM_DESC_ID: return InternalOrder.TERM_DESC;
-                case TERM_ASC_ID: return InternalOrder.TERM_ASC;
+                case COUNT_DESC_ID: return InternalOrder.COUNT_DESC;
+                case COUNT_ASC_ID: return InternalOrder.COUNT_ASC;
+                case KEY_DESC_ID: return InternalOrder.KEY_DESC;
+                case KEY_ASC_ID: return InternalOrder.KEY_ASC;
                 case Aggregation.ID:
                     boolean asc = in.readBoolean();
                     String key = in.readString();
                     return new InternalOrder.Aggregation(key, asc);
                 case CompoundOrder.ID:
                     int size = in.readVInt();
-                    List<Terms.Order> compoundOrder = new ArrayList<>(size);
+                    List<BucketOrder> compoundOrder = new ArrayList<>(size);
                     for (int i = 0; i < size; i++) {
-                        compoundOrder.add(Streams.readOrder(in, false));
+                        compoundOrder.add(Streams.readOrder(in));
+                    }
+                    return new CompoundOrder(compoundOrder, false);
+                default:
+                    throw new RuntimeException("unknown order");
+            }
+        }
+    }
+
+    public static class Parser {
+
+        public static BucketOrder parseOrderParam(XContentParser parser, QueryParseContext context) throws IOException {
+            XContentParser.Token token;
+            String orderKey = null;
+            boolean orderAsc = false;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    orderKey = parser.currentName();
+                } else if (token == XContentParser.Token.VALUE_STRING) {
+                    String dir = parser.text();
+                    if ("asc".equalsIgnoreCase(dir)) {
+                        orderAsc = true;
+                    } else if ("desc".equalsIgnoreCase(dir)) {
+                        orderAsc = false;
+                    } else {
+                        throw new ParsingException(parser.getTokenLocation(),
+                            "Unknown order direction [" + dir + "]");
                     }
-                    return new CompoundOrder(compoundOrder, absoluteOrder);
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(),
+                        "Unexpected token " + token + " for [order]");
+                }
+            }
+            if (orderKey == null) {
+                throw new ParsingException(parser.getTokenLocation(),
+                    "Must specify at least one field for [order]");
+            }
+            return resolveOrder(orderKey, orderAsc);
+        }
+
+        static BucketOrder resolveOrder(String key, boolean asc) {
+            switch(key) {
+                case "_term": // TODO deprecate
```

Comment:
Deprecation logging needs to be done using the `DeprecationLogger` class. If you look at `PutIndexTemplateRequest` you'll see an example. That should also add a warning header to the response.

Other things we need to do for deprecation include adding a deprecation note in the relevant place int he documentation, and adding the description of any breaking changes to the breaking changes document